ocrd-fileformat-transform

```diff
@@ -10,120 +11,95 @@ SHAREDIR="$(cd "$(dirname "$0")" >/dev/null && pwd)"
 SCRIPT_NAME="${0##*/}"
 
 MIMETYPE_PAGE=$(ocrd bashlib constants MIMETYPE_PAGE)
+NJOBS=2 # maximum number of concurrent processes (to better utilize CPU, since this is mostly I/O bound)
 
 main () {
     # Load ocrd bashlib functions
     # shellcheck source=../core/ocrd/bashlib/lib.bash
     source $(ocrd bashlib filename)
 
-    # Require at least 2.15.0
-    ocrd__minversion 2.15.0
-
     # Describe calling script to lib.bash
     ocrd__wrap "$SHAREDIR/ocrd-tool.json" "$SCRIPT_NAME" "$@"
+    ocrd__minversion 2.30.0
 
     cd "${ocrd__argv[working_dir]}"
     page_id=${ocrd__argv[page_id]:-}
     in_file_grp=${ocrd__argv[input_file_grp]}
     out_file_grp=${ocrd__argv[output_file_grp]}
     overwrite=${ocrd__argv[overwrite]}
-    mkdir -p $out_file_grp
-
+ 
     # read params
-    local from_to script_args output_extension
+    local from_to script_args out_extension out_mimetype
     script_args=(${params['script-args']:-})
     from_to=(${params['from-to']})
-    output_extension=${params['ext']}
-    if [[ -z "$output_extension" ]];then
+    out_extension=${params['ext']}
+    if [[ -z "$out_extension" ]];then
         case "${from_to[1]}" in
-            page*|alto*) output_extension=".xml" ;;
-            hocr*) output_extension=".html" ;;
-            text*) output_extension=".txt" ;;
+            page*|alto*) out_extension=".xml" ;;
+            hocr*) out_extension=".html" ;;
+            text*) out_extension=".txt" ;;
         esac
     fi
-
-    # Download the files and do the conversion
-    local IFS=$'\n'
-    files=($(ocrd workspace find \
-        ${page_id:+-g} ${page_id:-} \
-        -G $in_file_grp \
-        -k local_filename \
-        -k ID \
-        -k pageId \
-        -k mimetype \
-        --download))
-    local IFS=$' \t\n'
-    local n=0 zeros=0000
-    for csv in "${files[@]}"; do
-        let n+=1
-        # Parse comma separated fields
-        local IFS=$'\t'
-        local fields=($csv)
-        local IFS=$' \t\n'
-
-        local in_file="${fields[0]}"
-        local in_id="${fields[1]}"
-        local pageid="${fields[2]:-}"
-        local mimetype="${fields[3]:-}"
-
-        if [[ "$mimetype" =~ ^image/.* ]]; then
-            ocrd log -n "$SCRIPT_NAME" debug "ignoring image $in_id ($pageid)"
-            continue
-        elif ! test -f "$in_file"; then
-            ocrd log -n "$SCRIPT_NAME" error "input file \"$in_file\" ID=${in_id} (pageId=${pageid}) is not on disk"
-            continue
-        else
-            ocrd log -n "$SCRIPT_NAME" debug "processing input file $in_id ($pageid)"
+    case "${from_to[1]}" in
+        alto*) out_mimetype="application/alto+xml" ;;
+        page*) out_mimetype="$MIMETYPE_PAGE" ;;
+        hocr*) out_mimetype="text/html" ;;
+        text) out_mimetype="text/plain" ;;
+    esac
+
+    local results=$(mktemp -d)
+    for ((n=0; n<${#ocrd__files[*]}; n++)); do
+        if (($(jobs -p | wc -l) > $NJOBS + 1)); then
```

Comment:
That is a neat pattern. I ever only use the shell job controls in interactive mode, I was not even aware of `set -m` because I tend to use GNU parallel or xargs in such cases. Always nice to see what can be done in pure bash.